File: frontend/services/userService.ts
import apiClient from '@/lib/api-client';
import type {
  UserResponseDTO,
  CreateUserRequestDTO,
  SetDepartmentPayload,
  UpdateUserRequestDTO,
  UserActivityDTO,
  ActiveAdminCountDTO,
} from '@/types/user';
import type { Page } from '@/types/incident';

/** Current authenticated user profile. */
export async function getMe(): Promise<UserResponseDTO> {
  const { data } = await apiClient.get<UserResponseDTO>('/api/me');
  return data;
}

/** Paginated user list (ADMIN only). */
export async function getUsers(params: { page?: number; size?: number } = {}): Promise<Page<UserResponseDTO>> {
  const { data } = await apiClient.get<Page<UserResponseDTO>>('/api/users', {
    params: { page: params.page ?? 0, size: params.size ?? 50 },
  });
  return data;
}

/** Fetch a single user by id (ADMIN only). */
export async function getUser(id: string | number): Promise<UserResponseDTO> {
  const { data } = await apiClient.get<UserResponseDTO>(`/api/users/${id}`);
  return data;
}

/** Correct a user's identity (names only) — audit reason handled client-side. */
export async function updateUser(
  id: string | number,
  payload: UpdateUserRequestDTO,
): Promise<UserResponseDTO> {
  const { data } = await apiClient.put<UserResponseDTO>(`/api/users/${id}`, payload);
  return data;
}

/** Create a new user (ADMIN only). */
export async function createUser(payload: CreateUserRequestDTO): Promise<UserResponseDTO> {
  const { data } = await apiClient.post<UserResponseDTO>('/api/users', payload);
  return data;
}

/** Promote a SOUS_CHEF to CHEF_ATELIER (ADMIN only). */
export async function promoteUser(id: number): Promise<UserResponseDTO> {
  const { data } = await apiClient.put<UserResponseDTO>(`/api/users/${id}/promote`);
  return data;
}

/** Demote a CHEF_ATELIER back to SOUS_CHEF (resets password + department). */
export async function demoteUser(id: number): Promise<UserResponseDTO> {
  const { data } = await apiClient.put<UserResponseDTO>(`/api/users/${id}/demote`);
  return data;
}

/** Cancel a pending promotion for an unclaimed CHEF_ATELIER. */
export async function cancelPromotion(id: number): Promise<UserResponseDTO> {
  const { data } = await apiClient.put<UserResponseDTO>(`/api/users/${id}/cancel-promotion`);
  return data;
}

/** Reactivate a deactivated user (ADMIN only). */
export async function activateUser(id: number): Promise<UserResponseDTO> {
  const { data } = await apiClient.put<UserResponseDTO>(`/api/users/${id}/activate`);
  return data;
}

/** Deactivate a user (ADMIN only). */
export async function deactivateUser(id: number): Promise<UserResponseDTO> {
  const { data } = await apiClient.put<UserResponseDTO>(`/api/users/${id}/deactivate`);
  return data;
}

/** Number of active ADMIN accounts — last-admin guard (ADMIN only). */
export async function getActiveAdminCount(): Promise<ActiveAdminCountDTO> {
  const { data } = await apiClient.get<ActiveAdminCountDTO>('/api/users/active-admin-count');
  return data;
}

/**
 * On-demand activity analytics for a user (ADMIN only) — declared / claimed /
 * resolved counts plus per-day buckets. Computed server-side via SQL.
 */
export async function getUserActivity(id: number): Promise<UserActivityDTO> {
  const { data } = await apiClient.get<UserActivityDTO>(`/api/users/${id}/activity`);
  return data;
}

/** One-shot onboarding — assign the current user's department. */
export async function setMyDepartment(
  payload: SetDepartmentPayload,
): Promise<UserResponseDTO> {
  const { data } = await apiClient.patch<UserResponseDTO>(
    '/api/users/me/department',
    { departmentId: payload.departmentId },
  );
  return data;
}

File: backend/src/main/java/incident/management/system/service/UserService.java
package incident.management.system.service;

import incident.management.system.dto.CreateUserRequest;
import incident.management.system.dto.DepartmentResponse;
import incident.management.system.dto.UpdateUserRequest;
import incident.management.system.dto.UserActivityResponse;
import incident.management.system.dto.UserResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface UserService {

    UserResponse createUser(CreateUserRequest request);

    UserResponse updateUser(Long id, UpdateUserRequest request);

    UserResponse getUserById(Long id);

    UserResponse getUserByEmail(String email);

    UserResponse getUserByMatricule(int matricule);

    Page<UserResponse> getAllUsers(Pageable pageable);

    /**
     * On-demand activity analytics for the given user (declared / claimed /
     * resolved counts plus per-day buckets). Metrics are computed via SQL
     * {@code COUNT(*) + GROUP BY DATE(...)} — no denormalized counters.
     */
    UserActivityResponse getUserActivity(Long id);

    /**
     * Demotes a CHEF_ATELIER back to {@code SOUS_CHEF}: resets the password
     * (unclaimed sentinel — must re-claim if re-promoted) and clears the
     * department assignment.
     */
    UserResponse demoteToSousChef(Long id);

    /**
     * Cancels a pending promotion for an unclaimed CHEF_ATELIER: reverts the
     * role to {@code SOUS_CHEF}, clears unused password-reset tokens and
     * resets the account status to active.
     */
    UserResponse cancelPromotion(Long id);

    /**
     * Number of currently active ADMIN accounts — used by the last-active-admin
     * guard on the admin surface.
     */
    long countActiveAdmins();

    void deleteUser(Long id);

    UserResponse activateUser(Long id);

    UserResponse deactivateUser(Long id);

    //  Admin Department Subscriptions

    void subscribeToDepartment(Long adminId, Long departmentId);

    void unsubscribeFromDepartment(Long adminId, Long departmentId);

    List<DepartmentResponse> getSubscribedDepartments(Long adminId);

    /**
     * Promotes a SOUS_CHEF user to CHEF_ATELIER.
     * The promoted user will have an empty {@code passwordHash},
     * requiring them to claim their account on first login.
     */
    UserResponse promoteToChefAtelier(Long id);
}
